sql: TestTelemetryLogging failed

sql.TestTelemetryLogging [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6096786?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6096786?buildTab=artifacts#/) on master @ [6b7f0058325de8b8b321a08137d462c418c3047f](https://github.com/cockroachdb/cockroach/commits/6b7f0058325de8b8b321a08137d462c418c3047f):


```
=== RUN   TestTelemetryLogging
    test_log_scope.go:162: test logs captured to: /artifacts/tmp/_tmp/233e6aa31ff7ad29c896f0479cb5d0c4/logTestTelemetryLogging1398345658
=== CONT  TestTelemetryLogging
    telemetry_logging_test.go:482: -- test log scope end --
--- FAIL: TestTelemetryLogging (0.65s)
=== RUN   TestTelemetryLogging/select-*-limit-3-query
    telemetry_logging_test.go:349: select-*-limit-3-query: expected "SkippedQueries":3 found:
        {"Timestamp":1660550156089651084,"EventType":"sampled_query","Statement":"SELECT * FROM \"\".\"\".u LIMIT ‹3›","Tag":"SELECT","User":"root","ApplicationName":"telemetry-logging-test","ExecMode":"exec","NumRows":3,"Age":0.556264,"TxnCounter":21,"SkippedQueries":2,"CostEstimate":7.084999999999999,"Distribution":"full","PlanGist":"AgHSAQIAAQAAAgYC","SessionID":"170b75919b1e9bba0000000000000001","Database":"defaultdb","StatementID":"170b7591a971b6680000000000000001","TransactionID":"94433bf8-581b-4a92-90fc-9b7faffccb5f","StatementFingerprintID":4478901680921638675,"TotalScanRowsEstimate":3,"OutputRowsEstimate":3,"StatsAvailable":true,"NanosSinceStatsCollected":1537302174692405,"BytesRead":111,"RowsRead":3,"ContentionNanos":3}
    telemetry_logging_test.go:345: select-*-limit-3-query: expected no skipped queries, found:
        {"Timestamp":1660550156091062040,"EventType":"sampled_query","Statement":"SELECT * FROM \"\".\"\".u LIMIT ‹3›","Tag":"SELECT","User":"root","ApplicationName":"telemetry-logging-test","ExecMode":"exec","NumRows":3,"Age":0.517347,"TxnCounter":23,"SkippedQueries":1,"CostEstimate":7.084999999999999,"Distribution":"full","PlanGist":"AgHSAQIAAQAAAgYC","SessionID":"170b75919b1e9bba0000000000000001","Database":"defaultdb","StatementID":"170b7591a9874d080000000000000001","TransactionID":"6cd79031-4569-4d6d-98b2-59278190139a","StatementFingerprintID":4478901680921638675,"TotalScanRowsEstimate":3,"OutputRowsEstimate":3,"StatsAvailable":true,"NanosSinceStatsCollected":1537302176113577,"BytesRead":111,"RowsRead":3,"ContentionNanos":3}
    --- FAIL: TestTelemetryLogging/select-*-limit-3-query (0.00s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTelemetryLogging.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-18583

sql.TestTelemetryLogging [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6110289?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6110289?buildTab=artifacts#/) on master @ [b1bc50214dceb571c1d2b0b7ce5a179c4ace6ac1](https://github.com/cockroachdb/cockroach/commits/b1bc50214dceb571c1d2b0b7ce5a179c4ace6ac1):


```
=== RUN   TestTelemetryLogging
    test_log_scope.go:162: test logs captured to: /artifacts/tmp/_tmp/233e6aa31ff7ad29c896f0479cb5d0c4/logTestTelemetryLogging981519054
=== CONT  TestTelemetryLogging
    telemetry_logging_test.go:482: -- test log scope end --
--- FAIL: TestTelemetryLogging (0.60s)
=== RUN   TestTelemetryLogging/select-*-limit-2-query
    telemetry_logging_test.go:349: select-*-limit-2-query: expected "SkippedQueries":2 found:
        {"Timestamp":1660638613707124020,"EventType":"sampled_query","Statement":"SELECT * FROM \"\".\"\".u LIMIT ‹2›","Tag":"SELECT","User":"root","ApplicationName":"telemetry-logging-test","ExecMode":"exec","NumRows":2,"Age":0.561722,"TxnCounter":15,"CostEstimate":6.06,"Distribution":"full","PlanGist":"AgHSAQIAAQAAAgYC","SessionID":"170bc6053ed7153c0000000000000001","Database":"defaultdb","StatementID":"170bc6054e4fd96d0000000000000001","TransactionID":"2dc22563-f98a-4aea-bc4d-60165cbcc1af","StatementFingerprintID":4478901680921638675,"TotalScanRowsEstimate":2,"OutputRowsEstimate":2,"StatsAvailable":true,"NanosSinceStatsCollected":1625759792194718,"BytesRead":74,"RowsRead":2,"ContentionNanos":2}
    --- FAIL: TestTelemetryLogging/select-*-limit-2-query (0.00s)
=== RUN   TestTelemetryLogging/select-*-limit-3-query
    telemetry_logging_test.go:345: select-*-limit-3-query: expected no skipped queries, found:
        {"Timestamp":1660638613709483335,"EventType":"sampled_query","Statement":"SELECT * FROM \"\".\"\".u LIMIT ‹3›","Tag":"SELECT","User":"root","ApplicationName":"telemetry-logging-test","ExecMode":"exec","NumRows":3,"Age":0.753258,"TxnCounter":18,"SkippedQueries":2,"CostEstimate":7.084999999999999,"Distribution":"full","PlanGist":"AgHSAQIAAQAAAgYC","SessionID":"170bc6053ed7153c0000000000000001","Database":"defaultdb","StatementID":"170bc6054e73f71c0000000000000001","TransactionID":"f8ea40b5-a97e-4a01-b014-3071f663a4e3","StatementFingerprintID":4478901680921638675,"TotalScanRowsEstimate":3,"OutputRowsEstimate":3,"StatsAvailable":true,"NanosSinceStatsCollected":1625759794710652,"BytesRead":111,"RowsRead":3,"ContentionNanos":3}
    --- FAIL: TestTelemetryLogging/select-*-limit-3-query (0.00s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTelemetryLogging.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql.TestTelemetryLogging [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6139979?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6139979?buildTab=artifacts#/) on master @ [5c2c62ecf1bea60c807edc6b4da22d900ad4ae03](https://github.com/cockroachdb/cockroach/commits/5c2c62ecf1bea60c807edc6b4da22d900ad4ae03):


```
=== RUN   TestTelemetryLogging
    test_log_scope.go:162: test logs captured to: /artifacts/tmp/_tmp/233e6aa31ff7ad29c896f0479cb5d0c4/logTestTelemetryLogging2929113753
=== CONT  TestTelemetryLogging
    telemetry_logging_test.go:482: -- test log scope end --
--- FAIL: TestTelemetryLogging (0.65s)
=== RUN   TestTelemetryLogging/select-*-limit-2-query
    telemetry_logging_test.go:349: select-*-limit-2-query: expected "SkippedQueries":2 found:
        {"Timestamp":1660811532037131963,"EventType":"sampled_query","Statement":"SELECT * FROM \"\".\"\".u LIMIT ‹2›","Tag":"SELECT","User":"root","ApplicationName":"telemetry-logging-test","ExecMode":"exec","NumRows":2,"Age":0.794672,"TxnCounter":15,"CostEstimate":6.06,"Distribution":"full","PlanGist":"AgHSAQIAAQAAAgYC","SessionID":"170c6349ef94283b0000000000000001","Database":"defaultdb","StatementID":"170c6349fdf27c160000000000000001","TransactionID":"86e3a5fb-62ec-44d2-bd67-013989bb4dec","StatementFingerprintID":4478901680921638675,"TotalScanRowsEstimate":2,"OutputRowsEstimate":2,"StatsAvailable":true,"NanosSinceStatsCollected":1798678122217204,"BytesRead":74,"RowsRead":2,"ContentionNanos":2}
    --- FAIL: TestTelemetryLogging/select-*-limit-2-query (0.00s)
=== RUN   TestTelemetryLogging/select-*-limit-3-query
    telemetry_logging_test.go:345: select-*-limit-3-query: expected no skipped queries, found:
        {"Timestamp":1660811532043888211,"EventType":"sampled_query","Statement":"SELECT * FROM \"\".\"\".u LIMIT ‹3›","Tag":"SELECT","User":"root","ApplicationName":"telemetry-logging-test","ExecMode":"exec","NumRows":3,"Age":0.730298,"TxnCounter":18,"SkippedQueries":2,"CostEstimate":7.084999999999999,"Distribution":"full","PlanGist":"AgHSAQIAAQAAAgYC","SessionID":"170c6349ef94283b0000000000000001","Database":"defaultdb","StatementID":"170c6349fe59aa440000000000000001","TransactionID":"5a0a72c9-9261-4766-92c8-476506f05371","StatementFingerprintID":4478901680921638675,"TotalScanRowsEstimate":3,"OutputRowsEstimate":3,"StatsAvailable":true,"NanosSinceStatsCollected":1798678129087805,"BytesRead":111,"RowsRead":3,"ContentionNanos":3}
    --- FAIL: TestTelemetryLogging/select-*-limit-3-query (0.00s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTelemetryLogging.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

sql.TestTelemetryLogging [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6168815?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6168815?buildTab=artifacts#/) on master @ [a7fafb5ec450bb2a4be405d901081a7be515a5ff](https://github.com/cockroachdb/cockroach/commits/a7fafb5ec450bb2a4be405d901081a7be515a5ff):


```
=== RUN   TestTelemetryLogging
    test_log_scope.go:162: test logs captured to: /artifacts/tmp/_tmp/233e6aa31ff7ad29c896f0479cb5d0c4/logTestTelemetryLogging505735969
=== CONT  TestTelemetryLogging
    telemetry_logging_test.go:482: -- test log scope end --
--- FAIL: TestTelemetryLogging (0.68s)
=== RUN   TestTelemetryLogging/select-*-limit-2-query
    telemetry_logging_test.go:349: select-*-limit-2-query: expected "SkippedQueries":2 found:
        {"Timestamp":1660982760309434965,"EventType":"sampled_query","Statement":"SELECT * FROM \"\".\"\".u LIMIT ‹2›","Tag":"SELECT","User":"root","ApplicationName":"telemetry-logging-test","ExecMode":"exec","NumRows":2,"Age":0.656865,"TxnCounter":16,"SkippedQueries":1,"CostEstimate":6.06,"Distribution":"full","PlanGist":"AgHSAQIAAQAAAgYC","SessionID":"170cff05204a85b90000000000000001","Database":"defaultdb","StatementID":"170cff052e4b02260000000000000001","TransactionID":"01893783-5a33-4d7b-97e8-0e0fe046b529","StatementFingerprintID":4478901680921638675,"TotalScanRowsEstimate":2,"OutputRowsEstimate":2,"StatsAvailable":true,"NanosSinceStatsCollected":1969906394483282,"BytesRead":74,"RowsRead":2,"ContentionNanos":2}
    --- FAIL: TestTelemetryLogging/select-*-limit-2-query (0.00s)
=== RUN   TestTelemetryLogging/select-*-limit-3-query
    telemetry_logging_test.go:345: select-*-limit-3-query: expected no skipped queries, found:
        {"Timestamp":1660982760311301346,"EventType":"sampled_query","Statement":"SELECT * FROM \"\".\"\".u LIMIT ‹3›","Tag":"SELECT","User":"root","ApplicationName":"telemetry-logging-test","ExecMode":"exec","NumRows":3,"Age":0.934526,"TxnCounter":18,"SkippedQueries":1,"CostEstimate":7.084999999999999,"Distribution":"full","PlanGist":"AgHSAQIAAQAAAgYC","SessionID":"170cff05204a85b90000000000000001","Database":"defaultdb","StatementID":"170cff052e677a5c0000000000000001","TransactionID":"84879a6f-461b-4d9c-be5a-79a035e0c1d2","StatementFingerprintID":4478901680921638675,"TotalScanRowsEstimate":3,"OutputRowsEstimate":3,"StatsAvailable":true,"NanosSinceStatsCollected":1969906396443803,"BytesRead":111,"RowsRead":3,"ContentionNanos":3}
    --- FAIL: TestTelemetryLogging/select-*-limit-3-query (0.00s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTelemetryLogging.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/exec_log.go
+++ b/pkg/sql/exec_log.go
@@ -385,7 +385,7 @@ func (p *planner) maybeLogStatementInternal(
 		// We only log to the telemetry channel if enough time has elapsed from
 		// the last event emission.
 		requiredTimeElapsed := 1.0 / float64(maxEventFrequency)
-		_, tracingEnabled := p.curPlan.instrumentation.Tracing()
+		tracingEnabled := telemetryMetrics.isTracing(p.curPlan.instrumentation.Tracing())
 		// Always sample if the current statement is not of type DML or tracing
 		// is enabled for this statement.
 		if p.stmt.AST.StatementType() != tree.TypeDML || tracingEnabled {

--- a/pkg/sql/telemetry_logging.go
+++ b/pkg/sql/telemetry_logging.go
@@ -17,6 +17,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/settings"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 )
 
 // Default value used to designate the maximum frequency at which events
@@ -57,6 +58,9 @@ type TelemetryLoggingTestingKnobs struct {
 	// for the query. Used to stub non-zero values to populate the log's contention
 	// time field.
 	getContentionNanos func() int64
+	// getTracingStatus allows tests to override whether the current query has tracing
+	// enabled or not. Queries with tracing enabled are always sampled to telemetry.
+	getTracingStatus func() bool
 }
 
 // ModuleTestingKnobs implements base.ModuleTestingKnobs interface.
@@ -92,6 +96,13 @@ func (t *TelemetryLoggingMetrics) getContentionTime(contentionTimeInNanoseconds 
 		return t.Knobs.getContentionNanos()
 	}
 	return contentionTimeInNanoseconds
+}
+
+func (t *TelemetryLoggingMetrics) isTracing(_ *tracing.Span, tracingEnabled bool) bool {
+	if t.Knobs != nil && t.Knobs.getTracingStatus != nil {
+		return t.Knobs.getTracingStatus()
+	}
+	return tracingEnabled
 }
 
 func (t *TelemetryLoggingMetrics) resetSkippedQueryCount() (res uint64) {

--- a/pkg/sql/telemetry_logging_test.go
+++ b/pkg/sql/telemetry_logging_test.go
@@ -23,7 +23,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -68,11 +67,27 @@ func (s *stubQueryMetrics) ContentionNanos() int64 {
 	return s.contentionNanos
 }
 
+type stubTracingStatus struct {
+	syncutil.RWMutex
+	isTracing bool
+}
+
+func (s *stubTracingStatus) setTracingStatus(t bool) {
+	s.RWMutex.Lock()
+	defer s.RWMutex.Unlock()
+	s.isTracing = t
+}
+
+func (s *stubTracingStatus) TracingStatus() bool {
+	s.RWMutex.RLock()
+	defer s.RWMutex.RUnlock()
+	return s.isTracing
+}
+
 // TestTelemetryLogging verifies that telemetry events are logged to the telemetry log
 // and are sampled according to the configured sample rate.
 func TestTelemetryLogging(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 86118, "flaky test")
 	sc := log.ScopeWithoutShowLogs(t)
 	defer sc.Close(t)
 
@@ -81,6 +96,7 @@ func TestTelemetryLogging(t *testing.T) {
 
 	st := stubTime{}
 	sqm := stubQueryMetrics{}
+	sts := stubTracingStatus{}
 
 	s, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{
 		Knobs: base.TestingKnobs{
@@ -92,6 +108,7 @@ func TestTelemetryLogging(t *testing.T) {
 			TelemetryLoggingKnobs: &TelemetryLoggingTestingKnobs{
 				getTimeNow:         st.TimeNow,
 				getContentionNanos: sqm.ContentionNanos,
+				getTracingStatus:   sts.TracingStatus,
 			},
 		},
 	})
@@ -143,6 +160,7 @@ func TestTelemetryLogging(t *testing.T) {
 		expectedWrite           bool
 		expectedErr             string // Empty string means no error is expected.
 		contentionNanos         int64
+		enableTracing           bool
 	}{
 		{
 			// Test case with statement that is not of type DML.
@@ -163,6 +181,7 @@ func TestTelemetryLogging(t *testing.T) {
 			expectedRead:            false,
 			expectedWrite:           false,
 			contentionNanos:         0,
+			enableTracing:           false,
 		},
 		{
 			// Test case with statement that is of type DML.
@@ -181,6 +200,7 @@ func TestTelemetryLogging(t *testing.T) {
 			expectedRead:            false,
 			expectedWrite:           false,
 			contentionNanos:         1,
+			enableTracing:           false,
 		},
 		{
 			// Test case with statement that is of type DML.
@@ -200,6 +220,7 @@ func TestTelemetryLogging(t *testing.T) {
 			expectedRead:            true,
 			expectedWrite:           false,
 			contentionNanos:         2,
+			enableTracing:           false,
 		},
 		{
 			// Test case with statement that is of type DML.
@@ -218,6 +239,7 @@ func TestTelemetryLogging(t *testing.T) {
 			expectedRead:            true,
 			expectedWrite:           false,
 			contentionNanos:         3,
+			enableTracing:           false,
 		},
 		{
 			// Test case with a full scan.
@@ -236,6 +258,7 @@ func TestTelemetryLogging(t *testing.T) {
 			expectedRead:            true,
 			expectedWrite:           false,
 			contentionNanos:         0,
+			enableTracing:           false,
 		},
 		{
 			// Test case with a write.
@@ -254,6 +277,7 @@ func TestTelemetryLogging(t *testing.T) {
 			expectedRead:            true,
 			expectedWrite:           true,
 			contentionNanos:         0,
+			enableTracing:           false,
 		},
 		// Not of type DML so not sampled
 		{
@@ -271,6 +295,27 @@ func TestTelemetryLogging(t *testing.T) {
 			expectedRead:            false,
 			expectedWrite:           false,
 			expectedErr:             "a role/user named ‹root› already exists",
+			enableTracing:           false,
+		},
+		{
+			// Test case with statement that is of type DML.
+			// All statements should be logged despite not exceeding the necessary elapsed time (0.1s)
+			// due to tracing being enabled.
+			name:                    "select-with-tracing",
+			query:                   "SELECT * FROM u LIMIT 4;",
+			queryNoConstants:        "SELECT * FROM u LIMIT _",
+			execTimestampsSeconds:   []float64{10, 10.01, 10.02, 10.03, 10.04, 10.05},
+			expectedLogStatement:    `SELECT * FROM \"\".\"\".u LIMIT ‹4›`,
+			stubMaxEventFrequency:   10,
+			expectedSkipped:         []int{0, 0, 0, 0, 0, 0},
+			expectedUnredactedTags:  []string{"client"},
+			expectedApplicationName: "telemetry-logging-test",
+			expectedFullScan:        false,
+			expectedStatsAvailable:  true,
+			expectedRead:            true,
+			expectedWrite:           false,
+			contentionNanos:         2,
+			enableTracing:           true,
 		},
 	}
 
@@ -280,6 +325,7 @@ func TestTelemetryLogging(t *testing.T) {
 			stubTime := timeutil.FromUnixMicros(int64(execTimestamp * 1e6))
 			st.setTime(stubTime)
 			sqm.setContentionNanos(tc.contentionNanos)
+			sts.setTracingStatus(tc.enableTracing)
 			_, err := db.DB.ExecContext(context.Background(), tc.query)
 			if err != nil && tc.expectedErr == "" {
 				t.Errorf("unexpected error executing query `%s`: %v", tc.query, err)

